# Import Libraries

In [1]:
import tensorflow as tf
from tensorflow.keras import layers, models
from tensorflow.keras.datasets import mnist

# Load MNIST

In [2]:
(x_train, y_train), (x_test, y_test) = mnist.load_data()

# Reshape

In [3]:
x_train = x_train.reshape(-1, 28, 28, 1).astype("float32") / 255
x_test  = x_test.reshape(-1, 28, 28, 1).astype("float32") / 255

# Build CNN Model

In [4]:
model = models.Sequential([
    layers.Conv2D(32, (3,3), activation='relu', input_shape=(28,28,1)),
    layers.MaxPooling2D((2,2)),
    
    layers.Conv2D(64, (3,3), activation='relu'),
    layers.MaxPooling2D((2,2)),
    
    layers.Conv2D(64, (3,3), activation='relu'),

    layers.Flatten(),
    layers.Dense(128, activation='relu'),
    layers.Dense(10, activation='softmax')
])

d:\anaconda\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


# Compile

In [5]:
model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

# Train

In [6]:
history = model.fit(
    x_train, y_train,
    epochs=5,
    batch_size=64,
    validation_split=0.1
)

Epoch 1/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 16s 16ms/step - accuracy: 0.9451 - loss: 0.1788 - val_accuracy: 0.9830 - val_loss: 0.0564
Epoch 2/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 15s 17ms/step - accuracy: 0.9847 - loss: 0.0491 - val_accuracy: 0.9888 - val_loss: 0.0403
Epoch 3/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 11s 13ms/step - accuracy: 0.9891 - loss: 0.0347 - val_accuracy: 0.9892 - val_loss: 0.0362
Epoch 4/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 11s 13ms/step - accuracy: 0.9914 - loss: 0.0274 - val_accuracy: 0.9913 - val_loss: 0.0345
Epoch 5/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 11s 13ms/step - accuracy: 0.9931 - loss: 0.0219 - val_accuracy: 0.9917 - val_loss: 0.0363


# Evaluate

In [7]:
test_loss, test_acc = model.evaluate(x_test, y_test)
print("Test Accuracy:", test_acc)

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9909 - loss: 0.0309
Test Accuracy: 0.9908999800682068


The MNIST CNN practicum demonstrates outstanding results, with the model reaching 99.09% accuracy on both the training and test sets and achieving a low loss of 0.0309. This indicates that the network has learned to identify handwritten digits with remarkable precision. The minimal loss shows strong confidence in its predictions, while the nearly identical training and testing accuracy suggests excellent generalization with no signs of overfitting. Overall, these outcomes highlight that the CNN—built with convolutional, pooling, and dense layers—is highly effective for digit classification and successfully extracts the key features needed to differentiate all ten MNIST classes.
